DASH pipeline packet flow update proposal. (#449)
sonic-net/DASH: documentation/general/dash-sai-pipeline-packet-flow.md

![](./images/dash-data-path-overview.svg)

DASH-SAI APIs are designed to be generic, so we could use it to set up the data path in different ways as needed. 

- The simplest way to set up the data path is to use the DASH-SAI APIs to create your pipeline and offload all your policy to ASIC.
- If your application requires more advanced policy, then the techonology provider could also provide a inbox data plane app to implement these policies. And still, this is transparent to the upper layer, everything is hidden under the SAI level.
- If even more customized or complicated pipelines are needed for certain traffic, we could also provide our own customized data plane app and use the DASH policy to instruct the ASIC to trap these packets up when they shows up.

## 3. Packet Structure

Before diving into the pipeline and packet flow, to better describe the behaviors, let's first have a clear understanding on the packet structure that DASH supports.

Overall, the high-level packet structure looks like below:

- The inner most packet is whatever the customer sends, which is called overlay. 

> - The first encap is called underlay, which is the most frequently used layer for implement any virtual network functions, such as VNET routing, load balancer, etc.

Reviewer: Since the underlay also requires a tunnel encapsulation, wouldn't it be simple to model the layers as | ** | Tunnel 2 | Tunnel 1 | Tunnel 0 | Overlay | where Tunnel0 is used for underlay?

Author: that's actually true. however, this underlay concepts is already everywhere in today's DASH pipeline, such as mapping table, eni and etc.... so maybe better to keep it as it is for now.

but to allow this change for future, we could use | ** | Tunnel 2 | Tunnel 1 | Underlay | Overlay |. does this sound good to you?

Author: updated.

Reviewer: Yeah, that's much better. Now, Tunne0 and underlay can be used interchangeably, if needed.

```mermaid
flowchart TD
    subgraph ENI
        EL[ENI Lookup]

        EP0[ENI 0 Pipeline]
        EP1[ENI 1 Pipeline]
        EP2[ENI 2 Pipeline]
    end

    subgraph LoadBalancer
        VL[Vip Lookup]

        LBP0[LB 0 Pipeline]
        LBP1[LB 1 Pipeline]
        LBP2[LB 2 Pipeline]
    end

    EL --> |ENI 0| EP0
    EL --> |ENI 1| EP1
    EL --> |ENI 2| EP2

    VL --> |LB 0| LBP0
    VL --> |LB 1| LBP1
    VL --> |LB 2| LBP2
```

A pipeline can also define its initial matching stages, which will be used for starting processing the packets if no flow is matched.

> NOTE:
> 
> DASH-SAI pipeline is a logical concept. It doesn't have to be 1 to 1 mapping to a physical ASIC pipeline. The underlying implementation can be as simple as a metadata field update with a specific value, when the entry getting matched, e.g., `ENI = Inner Source/Destination MAC`.

### 5.4. Packet Decap

> If a pipeline is found, before processing the packets, all outer encaps will be decap'ed, and with key information saved in metadata bus, such as encap type, source IP and VNI / GRE Key, exposing the inner most packet going through the pipeline. This simplifies the flow matching logic and also allow us to create the reverse flow properly.

Reviewer 2: Can you clarify how to decap all outer encaps? I'm thinking we probably don't want to simply parse headers until there are no more tunnels found. What if the customer traffic is actually a tunnel they created?

Author: yea, I was thinking about adding these details but also afraid going too far on this XD. But here is my thoughts:

## Problem of preserving the encaps

Today, the encaps of incoming packet (before pipeline processing) will be used for creating the encaps of the reverse flow, so essentially, we have 2 things to answer:

1. How preserving encaps affects the reverse flow creation?
2. How preserving encaps affects the forwarding side tunnel creation?

With the packet flow doc, now we have a clear definition of the packet structure: overlay (inner most) -> underlay -> tunnel0 -> tunnel1 -> .... However, depends on what user wants, they could do something really interesting, for example, removing or updating the underlay, preserve the tunnel0 and remove tunnel1. So, we are facing all the combinations of CURD of all the encaps including structure change: after removing underlay, should tunnel0 becomes underlay or should we keep it as tunnel0? This affects how the `staticencap` and `tunnel` action works today.

## Solution

In my thought, since these are essentially 2 questions, then we can model it as 2 independent operations. 

### Reverse tunnel creation

For reverse tunnel creation, Since no matter which encap (nvgre or vxlan) we use, they always have a key inside, so we can build the VNI table to specify what we should do. This is already covered with "stateless decap and stateful decap" section. If a VNI is stateless, then we don't create the reverse encap, otherwise we create one.

```json
{
    "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "stateless": false },
    "DASH_SAI_VNI_TABLE|12346": { "direction": "outbound", "stateless": true }
}
```

If automatic reverse tunnel creation is not enough, we can also use `reverse_tunnel` to create a dedicated one.

### Forwarding tunnel creation

Although the encap is decaped, but all the encap information is preserved in the metadata bus. Hence, we can easily recreate them in anyway we want by introducing a new action, say: `tunnel_from_encap` (sorry, the name might look silly), and defined as below:

- Parameters from action:
  - Source: "underlay|tunnel0|tunnel1|..."
  - Target: "underlay|tunnel0|tunnel1|..."
- Parameters from metadata:
  - `tunnel_from_encap_(underlay|tunnel0|tunnel1)_sip`: overrides the source ip in the specified target encap
  - `tunnel_from_encap_(underlay|tunnel0|tunnel1)_dip`: overrides the destination ip in the specified target encap
- Action:
  - Enable the target encap by copying the information from the source encap.
  - When overrides exists, uses the overrided one.

If this makes sense, I will add them into the doc.

Author: Updated the doc according. Please feel free to take another look, Bud! : D

Reviewer 2: Hi Riff, you wrote "Today, the encaps of incoming packet (before pipeline processing) will be used for creating the encaps of the reverse flow".  I'm not sure if this is true with the dash behavior that's been defined so far. 

In regard to the double decap, I think you are saying run the pipeline twice? After the first tunnel header is decapped, run the pipeline and the LPM/actions etc will direct us to run the pipeline again (and thus trigger another decap)? If I have this wrong please correct me.

I have a higher-level comment sort of related to this and the rest of the doc. I see you are defining the dash pipeline to make it more generic and modular. But this is not the approach that has been taken up until now. Maybe the use cases for DASH have expanded? I'm a little bit concerned about the effort it will take to implement what you're proposing/defining.

Author: > Hi Riff, you wrote "Today, the encaps of incoming packet (before pipeline processing) will be used for creating the encaps of the reverse flow". I'm not sure if this is true with the dash behavior that's been defined so far.

Hi Bud, feel free to correct me, if my understanding is not right. 

I believe what you are referring to is this: "For the inbound direction, after LPM/ACL lookup, the pipeline uses the underlay_ip as specified in the ENI table to Vxlan encapsulate the packet.". 

In my understanding, this is right when no flow is created on the inbound side. But if we allow outbound direction creating a flow and let the flow to handle the inbound packet, then this underlay_ip will not be used anymore, since the transformation should be coming from the flow. This maps to how fast path in DPU/ASIC. So, we need to get the IP and stored it into the reverse flow (for inbound side).

The simple idea is to use the underlay_ip in reverse flow. But if source VM using multiple VTEP as ECMP group, it won't be handled well, because it is preferred to use whichever VTEP that sends the traffic to return the packet. Additionally, if the ENI is used for modeling other things which handles multiple sources, such as load balancer or other gateways, the source IP will not be a single IP or can be configured on the ENI. So it is better for us to learn it from the original packet.

> In regard to the double decap, I think you are saying run the pipeline twice? After the first tunnel header is decapped, run the pipeline and the LPM/actions etc will direct us to run the pipeline again (and thus trigger another decap)? If I have this wrong please correct me.

Not really. We don't recirculate the packet, but let the parser to handle certain level of the encaps. What you are saying is actually a very good one. I believe the confusion is caused by we didn't have the maximum number of encaps defined. If this solves your concern, I will update the doc.

> I have a higher-level comment sort of related to this and the rest of the doc. I see you are defining the dash pipeline to make it more generic and modular. But this is not the approach that has been taken up until now. Maybe the use cases for DASH have expanded? I'm a little bit concerned about the effort it will take to implement what you're proposing/defining.

What you are feeling is right. I think DASH is doing good job modeling certain VNET scenarios. And it has all the corner stones to be flexible and modular, but it is not using them at all. For example, routing action behavior changes based on routing types. Routing types is used as the real minimum building block, making routing action almost useless. These things limit the usage of DASH a lot.

If you look at the proposal and ignore the inbound pipeline and multi-stage update, you will see the object model is not changed much at all (VNI table is introduced for handling multi-layer encap and multi-VNI cases, so it doesn't really count). But you might already feel the pipeline being much more flexible and can do a lot of more things.

For your concern, yes, I share the same concern as well. There are certain fundamental things being changed in this proposal, such as moving from programming routing type to programming routing action. It could increase the underlying implementation complexity. But on the other hand, with a simple mind model, the complexity could be highly reduced as well. No more corner cases, no more special handling, everything just moves as you see it.

So, although I have the same concern, I still would like to get this doc and proposal out, discussed and gather the feedbacks. And after this, we will be able to have a clear definition on how the pipeline works instead of having many actions being vague and do whichever they like under different conditions.

Hope this solves your question!

Author: the maximum number of decap is added into the doc.

Reviewer 2: Hi Riff,
Just writing down my comment from the last meeting regarding handling packets that arrive with two tunnel encaps. IMO always decapping the two tunnel headers is not a good solution. If the outer tunnel header is for DASH and the inner tunnel is a customer tunnel, then we would probably drop the packet. (The customer tunnel wouldn't match any valid DASH tunnel.) This basically means that DASH doesn't let customers use tunnels, which is kind of a severe limitation.

Not sure what the solution is. I think ideally you'd like to check the outer tunnel to know if a second tunnel is expected. For instance, if a mux can put a second encap on on a packet, then maybe the VNI/tunnel_id the mux applies is known to DASH orchestration. And the DASH direction_lookup table (which has the tunnel_id for a key) includes a double_encap flag. This would be set to true for any tunnel from a mux. At packet forwarding time, we do the direction_table lookup. If this flag is not set in the lookup result then we decap only the outer tunnel. If the flag is set, then not only do we decap a second tunnel, but we drop the packet if a second tunnel is not present. I know there are issues with implementing this with the P4 parser -- I'm only suggesting a logical solution.

Author: Thanks a lot for capturing this Bud! I have updated the doc with multi-layer encap handling for this problem, which should be doable with P4 with `value_set` feature in parser. Do you mind to help take another look and see how it looks? : D

On high level, it works with 2 things:

1. use `final_encap` flag in VNI, so we can tell the parser that there is no more encap after this.
2. use some code to fix or extract the encap info in metadata after parsing, similar to what DASH is doing today.

#### Stateless decap vs stateful decap

During the direction lookup stage, all other encaps will be examined, such as VNI lookup. For each VNI or GRE key, we can specify whether it is stateless or stateful.

Although the encap information will still be saved in the metadata bus, however, for stateless decap, we will not create the encap for the reverse flow. Otherwise (for the stateful decap), we will create the encap for the reverse flow by reversing the source and destination information.

```json
{
    "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "stateless": false },
    "DASH_SAI_VNI_TABLE|12346": { "direction": "outbound", "stateless": true }
}
```

### 5.5. Conntrack Lookup and Update

After entering a specific pipeline, the first stage will be the Conntrack Lookup stage, which does the flow lookup. If any flow is matched, the saved actions will be applied, the metering counters will be updated, and the rest of pipeline will be skipped.

The core of the Conntrack Lookup and Update stage is the flow table.

#### 5.5.1. Flow lookup

First, let's define the flow lookup behavior.

> The flow lookup **MUST** use the information of inner most packet. And the matching keys can be configurated via the DASH flow APIs during the DASH initialization. This allows us to support different flow matching behaviors for certain cases. For example, in L3 level routing, we don't need to create 5 tuple flows.

Reviewer 2: If I understand correctly, you are not saying that arbitrary fields can be used as the flow keys. Instead, from the packet there is at most the classic 5-tuple fields. And any of these 5 fields can be disabled from the key based on a config given at DASH init time? (To disable the field basically means using a fixed value like 0).

Author: Hi Bud, thanks a lot for review and leaving comments! For your questions:

> If I understand correctly, you are not saying that arbitrary fields can be used as the flow keys. Instead, from the packet there is at most the classic 5-tuple fields.

Yes, modeling the arbitrary fields as flow key will be very complicated, and also hard to implement on certain DPU/ASICs. So, the classic 5 tuple besides certain frequently used fields should be a place to start with. More details can be found here in the flow api spec (in review): See `sai_dash_flow_key_t` in <https://github.com/sonic-net/SONiC/blob/d2bb1913dacf682cfa2affc0723101df5fc6ed81/doc/dash/dash-flow/DASH_FLOW_SAI.md>. 

> And any of these 5 fields can be disabled from the key based on a config given at DASH init time? (To disable the field basically means using a fixed value like 0).

Yes, and yes, using 0 is actually a valid way to implement this. There are certain DPUs that actually allow the stages to be dynamically created too, so the underlying implementation is flexible, depending on the architecture.

First of all, since we have multiple matching stages in the pipeline, we need a way to pass the information from the matched entries in the earlier stages to the later ones to help us making final decisions on packet transformation. And this is what metadata bus is for.

At high-level, the metadata bus is a set of fields that being carried all the way through the pipeline along with the packet. It contains:

- The information from the original packet, such as encap information.
- The information from each matched entry and related data structures, e.g., when a VNET mapping entry is matched, we will publish the information from VNET to the metadata bus. 
- The routing action and its inline parameters from the matched entry.

Implementation-wise, this is similar to Packet Header Vector or Bus in NPL.

### 5.2. Direction Lookup

In DASH-SAI pipeline, traffic are split into 2 directions: `inbound` and `outbound`. Each direction has its own pipeline (see pipeline overview above). When a new packet arrives, we will assign a direction to the packet, then process the packet in the corresponding pipeline. This ensures us to match the flow and transform the packet in the right way.

> 

Reviewer: For completeness, it would be nice to specify how the direction is determined. Is it always based on the VNI lookup on the 1st VXLAN header found? Can we assume that every packet will have one and only one vxlan header? I believe some of the DASH use cases have multiple NVGRE headers, but I don't remember seeing a header stack with multiple vxlan headers.

Author: Yes, multiple encaps could happen and we should be able to handle that. How to handle the multiple encap is already added in the "Packet Decap" section.

### 5.3. Pipeline Lookup

> DASH supports multi-tenancy model for traffic handling. A single device can have multiple pipelines, and each pipeline is used to handle traffic for a specific tenant. When a packet arrives, besides direction lookup, we also need pipeline lookup to determine which pipeline to use for processing the packet.

Reviewer: Nit: You may want to mention here that "each is pipeline is used to handle traffic for a specific tenant OR instance of a service (L4LB)"

Author: right! this is actually a good one. Since LB is also just an example, I have updated this part to " or whichever network function being modeled, such as a load balancer".

Author: updated now.

- If even more customized or complicated pipelines are needed for certain traffic, we could also provide our own customized data plane app and use the DASH policy to instruct the ASIC to trap these packets up when they shows up.

## 3. Packet Structure

Before diving into the pipeline and packet flow, to better describe the behaviors, let's first have a clear understanding on the packet structure that DASH supports.

Overall, the high-level packet structure looks like below:

- The inner most packet is the customer input, which is called overlay.
- The first encap is called underlay0, which is the most frequently used layer to implement any virtual network function, such as VNET routing, load balancer, etc.
- On top of underlay0, we can extend and have more layer of encaps, which can be used for implementing additional routing hops.

| ... (Outer most) | 2 | 1 | 0 (Inner most) |
| - | - | - | - | - |
| **...** | **Underlay 1** | **Underlay 0** | **Overlay** |

Today, DASH supports up to 2 layers of encaps: underlay0 and underlay1.

## 4. Pipeline Overview

> DASH-SAI pipeline is modeled as a list of stages. Each stage defines its own tables, and use the table entries to match packets per the processing logic inside the match-action pipeline (key, action id, action data), and publish the corresponding metadata when an entry is matched. After all stages are processed, a list of final routing actions will be defined. Then, by executing these routing actions, the packet will be transformed accordingly, and corresponding flows will be generated for to the direction of the packet.

Reviewer 3: ```suggestion
DASH-SAI pipeline is modeled as a list of stages. Each stage defines its own tables, and use the table entries to match packets per the processing logic inside the match-action pipeline (key, action id, action data), and publish the corresponding metadata when an entry is matched. After all stages are processed, a list of final routing actions will be defined. Then, by executing these routing actions, the packet will be transformed accordingly, and corresponding flows will be generated for the direction of the packet.
```

    CTU0 --> MS0
    MS0 --> Out

    CTL1 --> |Flow Miss| PreACL1
    CTL1 --> |Flow Hit| AA1
    PreACL1 --> MATS1
    MATS1 --> AA1
    AA1 --> |Flow Miss| PostACL1
    PostACL1 --> CTU1
    AA1 --> |Flow Hit| MS1
    AA1 --> |Flow Change| CTU1
    CTU1 --> MS1
    MS1 --> Out
```

This design allows our upper layers to be flexible and doesn't limit to any specific object models for modeling their own policies. For example, [SONiC-DASH pipeline](https://github.com/sonic-net/SONiC/blob/master/doc/dash/dash-sonic-hld.md#2-packet-flows) uses DASH-SAI pipeline to implement VNET routing scenarios by translating the [SONiC DASH APIs](https://github.com/sonic-net/sonic-dash-api/tree/master/proto) to DASH-SAI model:

- A DASH-SAI pipeline is used to represent a VM NIC (ENI).
- The VxLAN VNI is used to do the direction lookup.
- The inner MAC address is used for pipeline lookup, a.k.a. ENI lookup (ENI).
- Once it enters the corresponding DASH pipeline, the `outbound` pipeline will be used to process the packets sourced from the VM, while the `inbound` pipeline will be used to process the packets destined to the VM.

## 5. Pipeline components

### 5.1. Per-packet metadata bus

> First of all, since we have multiple matching stages in the pipeline, we need a way to pass the information from the matched entries in the earlier stages to the later ones to help us making final decisions on packet transformation. And this is what metadata bus is for.

Reviewer 3: ```suggestion
First of all, since we have multiple matching stages in the pipeline, we need a way to pass the information from the matched entries in the earlier stages to the later ones to help us make final decisions on packet transformation. And this is what metadata bus is for.
```

> At high-level, the metadata bus is a set of fields that being carried all the way through the pipeline along with the packet. It contains:

Reviewer 3: ```suggestion
At high-level, the metadata bus is a set of fields that are carried throughout the pipeline along with the packet. It contains:
```

- The information from the original packet, such as encap information.
- The information from each matched entry and related data structures, e.g., when a VNET mapping entry is matched, we will publish the information from VNET to the metadata bus.
- The routing action and its inline parameters from the matched entry.

Implementation-wise, this is similar to Packet Header Vector or Bus in NPL.

### 5.2. Packet parsing

> In ASIC, parser is usually the first stage of the pipeline. It parses the incoming packet and extract the information into metadata bus, such as the source and destination MAC, IP, etc. And based on these information, we will invoke the rest of pipeline.

Reviewer 3: ```suggestion
In an ASIC, the parser is usually the first stage of the pipeline. It parses the incoming packet and extracts the information into the metadata bus, such as the source and destination MAC, IP, etc. And based on this information, we will invoke the rest of pipeline.
```

#### 5.2.1. Multi-layer encap parsing

> Multi-layer encap brings the problem in the packet parsing, because of encap layer detection and overlay packet handling:

Reviewer 3: ```suggestion
Multi-layer encap generates a problem in the packet parsing, because of the encap layer detection and overlay packet handling:
```

> 1. In ASIC, parser is always parsing from the outermost bits to innermost bits. However, in DASH, the encap stack is defined reversely: underlay1 -> underlay0 -> overlay. Because the incoming packet could have various number of encaps, we need to be able to map the encaps to the right layer.

Reviewer 3: ```suggestion
1. In an ASIC, the parser is always parsing from the outermost bits to innermost bits. However, in DASH, the encap stack is defined reversely: underlay1 -> underlay0 -> overlay. Because the incoming packet could have various number of encaps, we need to be able to map the encaps to the right layer.
```

2. The overlay (customer) packet could be also using encaps, so we need to be able to tell which one is ours and which one is overlay.

> 3. The incoming packet could have various number of encaps, and each encap doesn't really know if the inner packet is a overlay or not. For example, from the SDN pipeline basic element doc, the [Inbound from LB](./sdn-pipeline-basic-elements.md#inbound-from-lb) and [Internal Load Balancer in VNET communication](./sdn-pipeline-basic-elements.md#internal-load-balancer-in-vnet-communication) actually shares the same outer most header.

Reviewer 3: ```suggestion
3. The incoming packet could have various number of encaps, and each encap is not aware whether the inner packet is a overlay or not. For example, from the SDN pipeline basic element doc, the [Inbound from LB](./sdn-pipeline-basic-elements.md#inbound-from-lb) and [Internal Load Balancer in VNET communication](./sdn-pipeline-basic-elements.md#internal-load-balancer-in-vnet-communication) actually shares the same outer-most header.
```

> To solve this problems, we use 2 things for encap handling:

Reviewer 3: ```suggestion
To solve this problems, we use 2 methods for encap handling:
```

> 1. We can specify if parser should stop parsing more encaps or not for VNI:

Reviewer 3: ```suggestion
1. We can specify if parser should stop parsing more encaps or not for the VNI:
```

   ```json
   { "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "final_encap": true } }
   ```

   When `final_encap` is set to true, the VNI will be sent to parser to force the parser to treat the next layer as overlay packet.

> 2. Since the parser doesn't know how many encaps beforehand, so parser can always start treating the outermost layer as underlay1. And whenever it sees an the protocol is not an encap or the VNI is unknown, it treat the packet as overlay packet. Then, after parsing is done, we will fix the encap information in metadata bus.

Reviewer 3: ```suggestion
2. Since the parser doesn't know how many encaps existed prior, the parser can always start treating the outermost layer as underlay1. And whenever it sees that the protocol is not an encap, or the VNI is unknown, it will treat the packet as overlay packet. Then, after parsing is complete, we will fix the encap information in metadata bus.
```

>    For example, in the [Inbound from LB](./sdn-pipeline-basic-elements.md#inbound-from-lb) case, the outer encap will start to be mapped to underlay1, the ethernet and IP part of the inner packet will be mapped into underlay0, while the TCP/UDP part will be mapped into overlay. Then, after parsing, we will extract the ethernet and IP parts in underlay0 as overlay, as well as extract the underlay1 as underlay0.

Reviewer 3: ```suggestion
   For example, in the [Inbound from LB](./sdn-pipeline-basic-elements.md#inbound-from-lb) case, the outer encap will initially be mapped to underlay1, the ethernet and IP part of the inner packet will be mapped into underlay0, while the TCP/UDP part will be mapped into overlay. Then, after parsing, we will extract the ethernet and IP parts in underlay0 as overlay, as well as extract the underlay1 as underlay0.
```

### 5.3. Direction Lookup

> In DASH-SAI pipeline, traffic are split into 2 directions: `inbound` and `outbound`. Each direction has its own pipeline (see pipeline overview above). When a new packet arrives, we will assign a direction to the packet, then process the packet in the corresponding pipeline. This ensures us to match the flow and transform the packet in the right way.

Reviewer 3: ```suggestion
In the DASH-SAI pipeline, traffic has 2 directions: `inbound` and `outbound`. Each direction has its own pipeline (see pipeline overview above). When a new packet arrives, we will assign a direction to the packet, then process the packet in the corresponding pipeline. This ensures we match the flow and transform the packet correctly.
```

### 5.4. Pipeline Lookup

> DASH supports multi-tenancy model for traffic handling. A single device can have multiple pipelines, and each pipeline is used to handle traffic for a specific tenant or whichever network function being modeled, such as a load balancer. When a packet arrives, besides direction lookup, we also need pipeline lookup to determine which pipeline to use for processing the packet.

Reviewer 3: ```suggestion
DASH supports a multi-tenancy model for traffic handling. A single device can have multiple pipelines, and each pipeline is used to handle traffic for a specific tenant, or the intended network function being modeled (such as a load balancer). When a packet arrives, besides direction lookup, we also need pipeline lookup to determine which pipeline to use for processing the packet.
```

> For example, if we like to implement a VM NIC with, then we can model one pipeline as one VM NIC, then use the inner packet MAC to find the pipeline. However, if we would like to implement a load balancer, we can use a DASH-SAI pipeline to represent a load balancer instance and use the Public IP to find the SAI pipeline.

Reviewer 3: ```suggestion
For example, if we want to implement a VM NIC with (uncertain what this means?  a VM NIC only?), then we can model one pipeline as one VM NIC, and use the inner packet MAC to find the pipeline. However, if we would like to implement a load balancer, we can use a DASH-SAI pipeline to represent a load balancer instance and use the Public IP to find the SAI pipeline.
```

Author: ouch! sorry, looks like some words are somehow lost here. i am adding them in with another commit.

```mermaid
flowchart TD
    subgraph ENI
        EL[ENI Lookup]

        EP0[ENI 0 Pipeline]
        EP1[ENI 1 Pipeline]
        EP2[ENI 2 Pipeline]
    end

    subgraph LoadBalancer
        VL[Vip Lookup]

        LBP0[LB 0 Pipeline]
        LBP1[LB 1 Pipeline]
        LBP2[LB 2 Pipeline]
    end

    EL --> |ENI 0| EP0
    EL --> |ENI 1| EP1
    EL --> |ENI 2| EP2

    VL --> |LB 0| LBP0
    VL --> |LB 1| LBP1
    VL --> |LB 2| LBP2
```

> A pipeline can also define its initial matching stages, which will be used for starting processing the packets if no flow is matched.

Reviewer 3: ```suggestion
A pipeline can also define its initial matching stages, which will be used to start processing the packets if no flow is matched.
```

> NOTE:
>

> > DASH-SAI pipeline is a logical concept. It doesn't have to be 1 to 1 mapping to a physical ASIC pipeline. The underlying implementation can be as simple as a metadata field update with a specific value, when the entry getting matched, e.g., `ENI = Inner Source/Destination MAC`.

Reviewer 3: ```suggestion
> The DASH-SAI pipeline is a logical concept. It doesn't have to be a 1 to 1 mapping to a physical ASIC pipeline. The underlying implementation can be as simple as a metadata field update with a specific value, when the entry is matched, e.g., `ENI = Inner Source/Destination MAC`.
```

### 5.5. Packet Decap

> If a pipeline is found, before processing the packets, all outer encaps will be decap'ed, and with key information saved in metadata bus, such as encap type, source IP and VNI / GRE Key, exposing the inner most packet going through the pipeline. This simplifies the flow matching logic and also allow us to create the reverse flow properly.

Reviewer 3: The first sentence is pretty long, is there any way to shorten it?

Author: sure thing!

#### 5.5.1. Multi-layer encap handling

> The incoming packet to DASH pipeline could have multiple layers of encap, however, the ASIC capacity is usually limited. The more encaps we handle in parser and decaps, the more the latency will be. Furthermore, using packet recirculation is also not ideal, because it can greatly reduces the throughput and increases the latency.

Reviewer 3: ```suggestion
The incoming packet to DASH pipeline could have multiple layers of encap, however, the ASIC capacity is usually limited. The more encaps we handle in the parser and decaps, adds latency.  Furthermore, using packet recirculation is also not ideal, because it can greatly reduce throughput and increase latency.
```

> The limit of having 2 layers of encap at maximum is trying to help this problem and avoid the pipeline having unexpected latency.

Reviewer 3: ```suggestion
The limit of having 2 layers of encap at maximum is an attempt to help this problem and avoid the pipeline potentially experiencing unexpected latency.
```

#### 5.5.2. Stateless decap vs stateful decap

During the direction lookup stage, all other encaps will be examined, such as VNI lookup. For each VNI or GRE key, we can specify whether it is stateless or stateful.

Although the encap information will still be saved in the metadata bus, however, for stateless decap, we will not create the encap for the reverse flow. Otherwise (for the stateful decap), we will create the encap for the reverse flow by reversing the source and destination information.

```json
{
    "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "stateless": false },
    "DASH_SAI_VNI_TABLE|12346": { "direction": "outbound", "stateless": true }
}
```

#### 5.5.3. Encap fields handling

##### 5.5.3.1. Handling DSCP

> DASH pipeline provides 2 modes for handling the DSCP: "Preserve model" and "Pipe model".

Reviewer 3: ```suggestion
The DASH pipeline provides 2 modes for handling the DSCP: "Preserve model" and "Pipe model".
```

Preserve model works very similar to `SAI_TUNNEL_DSCP_MODE_UNIFORM_MODEL` with a slight difference:

- The DSCP value of outer most encap header will be preserved.
- When any encap being added, the preserved DSCP value will be copied to the new encap header.
- The DSCP value of overlay packet (customer packet) shall never be modified or preserved and copied to the outer header.

On the other hand, pipe model works very similar to `SAI_TUNNEL_DSCP_MODE_PIPE_MODEL`:

- The DSCP value of existing encaps will not be preserved.
- When any encap being added, the DASH pipeline will specify the DSCP value, and all routing action will take this value and copy it to the encaps.

If no encaps are added, say, for traffic sending to Internet, the DSCP value of overlay packet will be exposed as it is, even-if original packet arrives with a non-zero value.

> This gives the cloud infra full control over the DSCP value and avoid the customer packet spoofing the fields.

Reviewer 3: ```suggestion
This gives the cloud infra full control over the DSCP value and will prevent the customer from packet spoofing the fields.
```

Author: i took your advice and updated this part slightly different. hope it makes sense.

- The DSCP value of outer most encap header will be preserved.
- When any encap being added, the preserved DSCP value will be copied to the new encap header.
- The DSCP value of overlay packet (customer packet) shall never be modified or preserved and copied to the outer header.

On the other hand, pipe model works very similar to `SAI_TUNNEL_DSCP_MODE_PIPE_MODEL`:

- The DSCP value of existing encaps will not be preserved.
- When any encap being added, the DASH pipeline will specify the DSCP value, and all routing action will take this value and copy it to the encaps.

If no encaps are added, say, for traffic sending to Internet, the DSCP value of overlay packet will be exposed as it is, even-if original packet arrives with a non-zero value.

This gives the cloud infra full control over the DSCP value and will prevent the customer from spoofing the fields.

##### 5.5.3.2. Handling TTL

TTL behavior for encap shall be "pipe" model (similar to SAI_TUNNEL_TTL_MODE_PIPE_MODEL):

- When adding encaps, TTL value shall be default set to 64.
- DASH pipeline shall not modify the TTL values in the overlay packet (customer packet).

#### 5.5.4. Encap preservation

> Sometimes, depends on the scenario to implement, the customer might want to preserve certain original encaps in the outgoing traffic. For example, say we receive a packet with structure: overlay -> underlay0 -> underlay1 -> underlay2. And we want to remove or update underlay0, preserve underlay1 and remove underlay2. This gives us the problem of handling all the CRUD combinations of all encaps, including structure changes: after removing underlay0, should underlay1 becomes underlay0 or should we keep it as underlay1? All these things affects the encap related routing actions and final packet we create.

Reviewer 3: ```suggestion
Sometimes, depending upon the scenario to implement, the customer might want to preserve certain original encaps in the outgoing traffic. For example, say we receive a packet with structure: overlay -> underlay0 -> underlay1 -> underlay2. And we want to remove or update underlay0, preserve underlay1 and remove underlay2. This gives us the problem of handling all the CRUD combinations of all encaps, including structure changes: after removing underlay0, should underlay1 becomes underlay0 or should we keep it as underlay1? All these questions affect the encap related routing actions and final packet we create.
```

> Since all the encap information is preserved in the metadata bus for flow creation anyway, to solve this problem, we can simply recreate them in anyway we want with routing action: `tunnel_from_encap`. It allows the source and target encap and their override value being set, which allows us to preserve the encaps in anyway we want and also ensures the clarity of final transformation that we are doing in the end.

Reviewer 3: ```suggestion
Since all of the encap information is preserved in the metadata bus for flow creation, to solve this problem, we can simply recreate them in the manner we want using the routing action: `tunnel_from_encap`. It allows the source and target encap and their override value to be set, which allows us to preserve the encaps exactly as intended, and also ensures the clarity of final transformation performed in the end.
```

Please note that: Encap preservation will not affect the reverse tunnel creation as "Stateless decap vs stateful decap" section described above. A encap can be preserved as well as used in reverse tunnel creation at the same time, since they are essentially 2 different things.

### 5.6. Conntrack Lookup and Update

> After entering a specific pipeline, the first stage will be the Conntrack Lookup stage, which does the flow lookup. If any flow is matched, the saved actions will be applied, the metering counters will be updated, and the rest of pipeline will be skipped.

Reviewer 3: ```suggestion
After entering a specific pipeline, the first stage will be the Conntrack Lookup stage, which performs the flow lookup. If any flow is matched, the saved actions will be applied, the metering counters will be updated, and the rest of pipeline will be exited.
```

The core of the Conntrack Lookup and Update stage is the flow table.

#### 5.6.1. Flow lookup

First, let's define the flow lookup behavior.

> The flow lookup **MUST** use the information of inner most packet. And the matching keys can be configurated via the DASH flow APIs during the DASH initialization. This allows us to support different flow matching behaviors for certain cases. For example, in L3 level routing, we don't need to create 5 tuple flows.

Reviewer 3: ```suggestion
The flow lookup **MUST** use the information of inner most packet. And the matching keys can be configurated via the DASH flow APIs during the DASH initialization. This allows us to support different flow matching behaviors for specific cases. For example, in L3 level routing, we don't need to create 5 tuple flows.
```

> After the flow lookup, if a flow is matched, we will apply the saved actions in the flow direction and skip the rest of the pipeline. Otherwise, we will continue the pipeline processing.

Reviewer 3: ```suggestion
After the flow lookup, if a flow is matched, we will apply the saved actions in the flow direction and exit the rest of the pipeline. Otherwise, we will continue the pipeline processing.
```

#### 5.6.2. Flow creation

For new connections, after all packet transformations are applied, we will create a new flow in the flow table.

> Since a connection can be bi-directional, the flows needs to be created as flow pairs: the one in the same direction as the packet is usually called forwarding flow, while the other direction is called reverse flow. Of course, the flow pairs **MUST** be stored based on the pipeline direction.

Reviewer 3: ```suggestion
Since a connection can be bi-directional, the flows need to be created as flow pairs: the one in the same direction as the packet is usually called forwarding flow, while the other direction is called reverse flow. Of course, the flow pairs **MUST** be stored based on the pipeline direction.
```

> - When outbound pipeline creates a flow, the forwarding flow should be created on the outbound side, while reverse flow should be created on the inbound side.

Reviewer 3: ```suggestion
- When the outbound pipeline creates a flow, the forwarding flow should be created on the outbound side, while the reverse flow should be created on the inbound side.
```

> - When inbound pipeline creates a flow, it creates the flow in reversed way.

Reviewer 3: ```suggestion
- When the inbound pipeline creates a flow, it creates the flow in reverse way.
```

And here is how to create the flow pairs:

- The forwarding flow creation is straight forward, because all the final transformations are defined in the flow actions and metadata bus.
- To properly create the reverse flow, we will learn the information from the original tunnels, and reverse them as return encaps, which is why we need to save all the original tunnel information when doing packet decaps.

##### 5.6.2.1. Tunnel learning

> You might already notice that, although we uses the outer encap from the original packet to create the reverse flow, but the outer encaps are decap'ed and not used in flow matching. This creates a problem when source side fails over. Like the graph shows below, the traffic shifted from VTEP 1 (Green side) to VTEP 2 (Blue side), which changes the source IP in the encap. If we don't do anything, the existing reverse flow will continue to send the packet back to green side and causing the traffic being dropped.

Reviewer 3: ```suggestion
You might already notice that, although we uses the outer encap from the original packet to create the reverse flow, but the outer encaps are decap'ed and not used in flow matching. This creates a problem when source side fails over. Like the graph shows below, the traffic shifted from VTEP 1 (Green side) to VTEP 2 (Blue side), which changes the source IP in the encap. If we don't do anything (such as save the information), the existing reverse flow will continue to send the packet back to green side and cause the traffic to be dropped.
```

![Tunnel learning](./images/dash-tunnel-learning.svg)

To address this problem, the source information of each encap **MUST** be saved in the flow and compared during flow match. These information should no be part of the key, but whenever they change, we should reprocess the packet and update the saved reverse tunnel in the existing flow.

##### 5.6.2.2. Asymmetrical encap handling

> There could be cases where the encaps are asymmetrical, which means the incoming packet and return packet uses different encaps. For example, in the case below, the encap of step 2 and 4 are not symmetrical. This means we could never recreate the flow on the DASH pipeline 2 side to tunnel the packet into the extra hop first with our current flow creation logic.

Reviewer 3: ```suggestion
There could be cases where the encaps are asymmetrical, which means the incoming packet and return packet use different encaps. For example, in the case below, the encap of step 2 and 4 are not symmetrical. This means we could never recreate the flow on the DASH pipeline 2 side to tunnel the packet into the extra hop first with our current flow creation logic.
```

![Asymmetrical encap](./images/dash-asymmetrical-encap.svg)

To fix this issue, a special action called `reverse_tunnel` is defined, which enables reverse side of the encaps.

#### 5.6.3. Flow resimulation

> When certain policies are changed, we might need to update the flows that is associated with the policies. For example, when a VM is moved to another place, the VNET mapping entry will be changed, and we need to update the existing flows that is associated with the VNET mapping entry. Otherwise, the outgoing traffic will be tunneled to incorrect place. This is called flow resimulation.

Reviewer 3: ```suggestion
In the event of a policy change, we might need to update the flows associated with the policies. For example, when a VM is moved to another location, the VNET mapping entry will be changed, and we need to update the existing flows associated with the VNET mapping entry. Otherwise, the outgoing traffic will be tunneled to an incorrect location. This is called flow resimulation.
```

> However, the flow resimulation is not as simple as removing the flow and let the packet to going through the pipeline again. For example:

Reviewer 3: ```suggestion
However, the flow resimulation is not as simple as removing the flow and sending the packet through the pipeline again. For example:
```

- When a ACL is changed in the outbound side, all flows needs to be resimulated. However, this cannot be implemented by removing all flow entries and let all flows to be recreated with the latest policy. The reason is that ACLs can be asymmetric and flows are created in pair. If the flows are removed, the inbound side traffic will be dropped immediately, because there is no inbound flow to make it bypass the ACLs.

> - Certain actions might need to bypass the flow resimulation to maintain the per flow consistency. For example, in load balancer case, we will want to ensure that flow resimulation will not causing us to forward existing connections to a different backend server.

Reviewer 3: ```suggestion
- Certain actions might need to bypass the flow resimulation to maintain the per flow consistency. For example, in a load balancer case, we will want to ensure that flow resimulation will not cause the pipeline to forward existing connections to a different backend server.
```

These requires us to implement the flow resimulation in a more sophisticated way, which is not fully modeled in DASH today. But we will come back to this design later.

### 5.7. Pre-pipeline ACL and Post-pipeline ACL

Pre-pipeline ACL and Post-pipeline ACL are used to drop the unexpected traffic before and after the packet transformation. It works as below:

> 1. As the high-level pipeline shows above, if an incoming packet hits a flow, it will skip matching all the ACLs. If an incoming packet is denied by a ACL, the packet will be dropped without creating a flow.

Reviewer 3: ```suggestion
1. As the high-level pipeline shows above, if an incoming packet hits a flow (i.e. is in the FastPath), it will skip matching all the ACLs. If an incoming packet is denied by a ACL, the packet will be dropped without creating a flow.
```

> 2. Both outbound and inbound pipeline has its own ACL stages, and only used for matching the packets in their direction.

Reviewer 3: ```suggestion
2. Both the outbound and inbound pipeline have their own ACL stages, and are used for matching the packets in their own direction.
```

    end

    subgraph Inbound Pipeline
        In_PreACL[Underlay ACL<br># Inbound Pre-Pipeline ACL #]
        In_MS[Match Stages]
        In_PostACL[Overlay ACL<br># Inbound Post-Pipeline ACL #]
    end

    Out_PreACL --> Out_MS
    Out_MS --> Out_PostACL

    In_PreACL --> In_MS
    In_MS --> In_PostACL
```

### 5.8. Routing actions and routing types

#### 5.8.1. Routing action

In DASH-SAI pipeline, routing actions are the fundamental building blocks for packet transformations. Each routing action is designed to work as below:

1. Take a specific list of metadata fields as input parameters:
   1. For example, `staticencap` action will take the `underlay_dip`, `underlay_sip`, `encap_type` and `encap_key` to encapsulate the packet.
   2. The parameters can come from 2 places - the metadata defined associated with the entries in each table, or the routing action definition itself. More details will be discussed in the next section.
2. Transform the packet in a specific way, e.g., encapsulate the packet, NAT'ing the address and port, etc.
3. Independent to other actions.
   1. With this design, we don't have to worry about the order of the actions.

>    2. This also enables the hardware to improve the E2E pipeline latency by doing the actions in parallel.

Reviewer 3: ```suggestion
   2. This also enables the hardware to improve the E2E pipeline latency by executing the actions in parallel.
```

| Stage type | Match Type | Match Fields | Metadata Behavior |
| ---------- | ---------- | ------------ | -------- |
| Routing | LPM | Source IP or Destination IP | Publish metadata from matched routing entry |
| IP Mapping | Exact Match | Source IP or Destination IP | Publish metadata from matched VNET and VNET mapping entry |
| Port Mapping | Range Match | Source Port + Destination Port | Publish metadata from matched port mapping entry |

For more on the metadata publishing, please refer to the metadata publishing section below.

#### 5.9.2. Pipeline profile and stage connections

Ideally, when DASH initializes or whenever we create a new pipeline, by simply creating multiple numbers of different types of matching stages and connecting them in different ways, we can easily implement different pipeline and network functions.

| Stage 0 | Stage 1 | Stage 2 | Stage 3 | Stage 4 |
| ------- | ------- | ------- | ------- | ------- |
| Routing 0 | IP Mapping 0 | Port Mapping | X | X |
| Routing 0 | Routing 1 | IP Mapping 0 | IP Mapping 1 | Port Mapping |
| Routing 0 | IP Mapping 0 | Port Mapping | Routing 1 | IP Mapping 1 |

> However, in reality, it might make the pipeline hard to implement, model, debug and validate or test at this moment. For example, changing matching fields, matching type and connection dynamically on pipeline creation is simply beyond the ability of P4, which is used by our behavior model today as well as many vendor's ASIC SDKs.

Reviewer 3: Do we want to mention P4 annotations that may help P4 bmv2?

Author: p4 annotations will help generating the sai headers, but sadly, it won't be able to help much in this case. :(

    Map1[IP Mapping 1]
    TcpPortMap[TCP Port Mapping]
    UdpPortMap[UDP Port Mapping]

    Routing0 --> | lpmrouting - 1 | Routing1
    Routing1 --> | maprouting - 0 | Map0
    Map0 --> | maprouting - 1 | Map1
    Map1 --> | portmaprouting | TcpPortMap
    Map1 --> | portmaprouting | UdpPortMap
```

#### 5.9.3. Stage transitions

##### 5.9.3.1. Stage transition routing type

To transit between stages, we use `transition` field to specify the transition routing type.

A transition routing type is a special type of [Routing Type](#572-routing-type):

1. Each transition routing type can only have one single routing action.
2. Only a special list of routing actions will take effect in the routing types, which are all related to stage transitions, as listed below.
3. If `transition` field is not set, we consider the packet is reaching the end of the pipeline and start executing the final routing actions.

| `action_type` value | Stage |
| ---------------- | ----- |
| `drop` | Drop the packet |
| `trap` | Sending it to CPU |
| `lpmrouting` | Skip to next routing stage |
| `maprouting` | Skip to next mapping stage |
| `portmaprouting` | Skip to next TCP or UDP Port Mapping stage |

> To help us being flexible, the transition routing actions can take a few parameters:

Reviewer 3: ```suggestion
To help us to stay flexible, the transition routing actions can take a few parameters:
```

- All transition routing actions other than `drop` and `trap`:
  - `default_routing_type`: If no entry is found, use this routing type to route the packet. If `default_routing_type` is not set, the packet will be dropped by default.
  - `stage_index`: If there are multiple stage of the same stage type, use this field to specify which stage to transit to. The index is starting from 0 and by default 0.
- IP-based routing actions (`lpmrouting` and `maprouting`):
  - `use_src_ip`: Use source IP in routing and mapping stages.
  - `ip_mask`: IP mask to apply before matching the entries.

Here is an example that shows how the routing stage entry looks like:

```json
// Routing stage entry:
// When this entry is matched, we will move to VNET mapping stage, which executes maprouting action and jump to VNET mapping stage.
"DASH_SAI_ROUTE_TABLE|123456789012|0|10.0.1.0/24": {
    "transition": "vnetmap",
    "vnet": "Vnet1"
}

"DASH_SAI_ROUTING_TYPE_TABLE|vnetmap": [
    {
        "action_type": "maprouting",
        "stage_index": 0
    }
]
```

##### 5.9.3.2. Stage skipping

> To avoid loop shows up in the pipeline, the pipeline is designed to be forward only. When transition routing action is invoked, we can simply update the target stage in the metadata, then the transition behavior can simply be described by the code below:

Reviewer 3: ```suggestion
To avoid loops appearing in the pipeline, the pipeline is designed to be forward only. When transition routing action is invoked, we can simply update the target stage in the metadata, then the transition behavior can simply be described by the code below:
```
